ServiceScan.SourceGenerator/Model/TypeModel.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace ServiceScan.SourceGenerator.Model;
+
+sealed record TypeModel
+{
+    public required string AssemblyName { get; init; }
+
+    public required bool IsAbstract { get; init; }
+
+    public required bool IsStatic { get; init; }
+
+    public required bool CanBeReferencedByName { get; init; }
+
+    public required TypeKind TypeKind { get; init; }
+
+    public required string DisplayString { get; init; }
+
+    public required string UnboundGenericDisplayString { get; init; }
+
+    public required bool IsGenericType { get; init; }
+
+    public required bool IsUnboundGenericType { get; init; }
+
+    public required EquatableArray<TypeModel> AllInterfaces { get; init; }
+
+    // OriginalDefinition is null for unbound generic types (to avoid recursion). For bound generic types, OriginalDefinition is the TypeModel representing the unbound generic type.
+    // For non-generic types, OriginalDefinition is null.
+    public required TypeModel? OriginalDefinition { get; init; }
+
+    public required TypeModel? BaseType { get; init; }
+
+    public static TypeModel? Create(SyntaxNode node, SemanticModel semanticModel, CancellationToken cancel)
+    {
+        if (node is not TypeDeclarationSyntax typeDeclaration)
+            return null;
+
+        if (semanticModel.GetDeclaredSymbol(typeDeclaration, cancel) is not INamedTypeSymbol symbol)
+            return null;
+
+        return Create(symbol);
+    }
+
+    public static TypeModel Create(INamedTypeSymbol symbol)
+    {
```

Comment:
room for improvement here:

even though there are only about 2k total types (mostly coming from System.Private.CoreLib), each of those TypeModels will have nested TypeModels for basetype and interfaces etc. Those instances are currently duplicated

Like if 2 unrelated types are constructed with Create, they will each construct their own BaseType TypeModel for `object`. Could attempt to de-duplicate references here if needed, use a cache of sorts

Reply:
OK I added a cache. This cache is keyed on the INamedTypeSymbol, which can't be kept across different compilations. So i make sure to only use it within the context of a single compilation. It doesn't affect the logic/accuracy - it is just a perf boost to keep the numbed of duplicate TypeModels down

So in the `FindServicesToRegister` method when we have the fully realized `ImmutableArray<TypeModel> refTypes`, it has ~2k top-level instances, before and after this change. However, before this change, VS memory snapshot showed 88k type model instances being held. After this change, that number is down to 3k

Reply:
Don't really like the idea of caching all the types available.
I'm thinking it probably makes sense to flip those steps over.
Get method data first, and combine it with SyntaxProvider and ReferencesProvider afterwards.
E.g. 
1. Get methods with attributes.

2A. Filter attributes with a AssemblyFrom value.
2B. Filter attributes without a AssemblyFrom value.

3A. Combine 2A with ReferenceProvider, get all types from corresponding reference matching attributes criteria, return AssemblyRegistration models.
3B. Combine 2B with SyntaxProvider, get all types matching attributes criteria, return AssemblyRegistration models.

4. Combine 3A with 3B, produce an output.

I'll try it out later, unless you do it first :)

Reply:
Ah yeah that sounds like a good plan - I'll leave that one to you!